block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -1034,6 +1082,121 @@ object BlockDagFileStorage {
             }
     } yield LMDBStore[F](env, dbi)
 
+  private def readBlockHashesByDeploy[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F]
+  ): F[List[(DeployId, BlockHash)]] = {
+    def readRec(
+        result: List[(DeployId, BlockHash)]
+    ): F[List[(DeployId, BlockHash)]] =
+      for {
+        deployIdSizeOpt <- randomAccessIO.readInt
+        result <- deployIdSizeOpt match {
+                   case Some(deployIdSize) =>
+                     val blockHashBytes = Array.ofDim[Byte](BlockHash.Length)
+                     val deployIdBytes  = Array.ofDim[Byte](deployIdSize)
+                     for {
+                       readDeployId  <- randomAccessIO.readFully(deployIdBytes)
+                       readBlockHash <- randomAccessIO.readFully(blockHashBytes)
+                       result <- (readDeployId, readBlockHash) match {
+                                  case (Some(()), Some(())) =>
+                                    readRec(
+                                      (ByteString.copyFrom(deployIdBytes) -> ByteString
+                                        .copyFrom(blockHashBytes)) :: result
+                                    )
+                                  case _ =>
+                                    Sync[F].raiseError[List[(DeployId, BlockHash)]](
+                                      BlockHashesByDeployLogIsCorrupted
+                                    )
+                                }
+                     } yield result
+                   case None =>
+                     result.reverse.pure[F]
+                 }
+      } yield result
+    readRec(List.empty)
+  }
+
+  def calculateBlockHashesByDeployCrc[F[_]: Monad](
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[Crc32[F]] = {
+    val crc = Crc32.empty[F]()
+    blockHashesByDeploy.traverse_[F, Unit] {
+      case (deployId, blockHash) =>
+        crc.update(deployId.size.toByteString.concat(deployId).concat(blockHash).toByteArray)
+    } >> crc.pure[F]
+  }
+
+  private def truncateBlockHashesByDeployLog[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F],
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[Unit] = {
+    val lastRecordSize: Long = BlockHash.Length.toLong + BlockHash.Length.toLong
+    for {
+      length <- randomAccessIO.length
+      _      <- randomAccessIO.setLength(length - lastRecordSize)
+    } yield ()
+  }
+
+  private def validateBlockHashesByDeploy[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F],
+      readBlockHashesByDeployCrc: Long,
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[(List[(DeployId, BlockHash)], Crc32[F])] =
+    for {
+      fullCalculatedCrc <- calculateBlockHashesByDeployCrc[F](blockHashesByDeploy)
+      result <- Monad[F].ifM(fullCalculatedCrc.value.map(_ == readBlockHashesByDeployCrc))(
+                 (blockHashesByDeploy, fullCalculatedCrc).pure[F],
+                 blockHashesByDeploy match {
+                   case Nil =>
```

Comment:
It reads better to have ```if (blockHashesByDeploy.isEmpty) ... else ...``` instead of ```case blockHashesByDeploy match Nil => ... case _ : _ => ..```